File: src/org/cerion/symcalc/expression/VarExpr.java
package org.cerion.symcalc.expression;

public class VarExpr extends Expr {

    public VarExpr(String str) {
        setValue(str);
    }

    @Override
	public ExprType getType() {
		return ExprType.VARIABLE;
	}

    @Override
    public void show(int i) {
    	indent(i,"VarExpr = " + value());
    }
    
    public String value() {
    	return (String)getValue();
    }
    
    @Override
    public String toString() {
    	return value();
	}
    
    @Override
    protected Expr evaluate() {
    	Expr result = getEnv().getVar(value());
    	if(result == null)
    		result = this;
    	
    	return result;
    }

    @Override
    public boolean equals(Expr e) {
        if(e.isVariable()) {
            VarExpr t = (VarExpr)e;
            if (value().contentEquals(t.value()))
                return true;
        }
    	
    	return false;
    }
}

File: src/org/cerion/symcalc/expression/VarExpr.kt
package org.cerion.symcalc.expression

class VarExpr(str: String) : Expr() {

    override val type: ExprType
        get() = ExprType.VARIABLE

    init {
        value = str
    }

    override fun show(i: Int) = indent(i, "VarExpr = " + value())
    override fun toString(): String = value()

    fun value(): String {
        return value as String
    }

    override fun evaluate(): Expr {
        var result = env.getVar(value())
        if (result == null)
            result = this

        return result
    }

    override fun equals(e: Expr): Boolean {
        if (e.isVariable) {
            val t = e as VarExpr
            if (value().contentEquals(t.value()))
                return true
        }

        return false
    }
}